CDAP-5424 Record lineage for external sources/sinks

poornachandra opened the pull request:
JIRA - https://issues.cask.co/browse/CDAP-5424

TODO:
- [ ] Identify input/output providers that are regular datasets - https://issues.cask.co/browse/CDAP-5675
- [x] Distinguish between name and alias for external datasets
- [x] Backwards compatibility
- [x] Add test
- [x] Use dataset runtime arguments to pass input/output configuration instead of dataset properties.

Build - http://builds.cask.co/browse/CDAP-DUT3961-15
Depends on https://github.com/caskdata/cdap/pull/5586

bdmogal commented:
@poornachandra I remember discussing a few months ago that we may need to track lineage of data coming into CDAP from an external source (e.g. data getting loaded into a stream from an external source). Can the concept in this PR be extended to that in future?

anwar6953 commented on cdap-app-templates/cdap-etl/cdap-etl-batch/src/main/java/co/cask/cdap/etl/batch/spark/SparkBatchSinkFactory.java:
```diff
@@ -156,16 +166,28 @@ private void addStageOutput(String stageName, String outputName) {
   }
 
-  private static final class BasicOutputFormatProvider implements OutputFormatProvider {
+  private static final class NamedOutputFormatProvider implements OutputFormatProvider {
```
this is sort of similar to `Input` class, then?

poornachandra replied:
This class is used for serialization/deserialization.

anwar6953 commented:
@bdmogal Isn't this PR fulfilling that exact requirement.

anew commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/BasicMapReduceContext.java:
```diff
@@ -237,6 +240,22 @@ private void addInput(String inputName, InputFormatProvider inputFormatProvider,
+    // Record lineage only if the input format has a proper name
+    if (!inputName.equals(inputFormatProvider.getInputFormatClassName())) {
+      try {
+        // Create an external dataset for the input format for lineage tracking
```
not sure I understand this. If I call addInput("one", fileSet1, null) and fileSet1 is a regular dataset (a FileSet), then you will create an externalDataset named "one-input"? So you create a "shadow" external dataset for every dataset is used as input in MR? Seems that something is wrong

poornachandra replied:
You are right, I'll remove the suffix from dataset name. Having a suffix is not necessary if we use datgaset runtime args to pass configuration around.

poornachandra replied:
Also, would it be sufficient to check if the InputFormatProvider is an instance of Dataset prevent creation an external Dataset for a regular Dataset?

anew commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/BasicMapReduceContext.java:
```diff
@@ -237,6 +240,22 @@ private void addInput(String inputName, InputFormatProvider inputFormatProvider,
+        // Create an external dataset for the input format for lineage tracking
+        // TODO: convert string to id
+        String datasetName = inputName + "-input";
+        getAdmin().createDataset(datasetName, "externalDataset",
```
does createDataset also record lineage? In that case you would generate two lineage records here

poornachandra replied:
No, create dataset does not record lineage.

anew commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/BasicMapReduceContext.java:
```diff
@@ -291,6 +310,19 @@ public void addOutput(String datasetName, Map<String, String> arguments) {
 
   @Override
   public void addOutput(String outputName, OutputFormatProvider outputFormatProvider) {
+    try {
```
same as for input, here you create an externalDataset for every existing fileset that is used as output in MR.

anew replied:
If you suffix sources with -input and sinks with -output, then how will the lineage be tracked if one pipeline writes to sink "xyz" and another pipeline reads from source "xyz"? Isn't the intention that this sink and source will be represented by the same external dataset? If you use different suffixes, then they will not.

anew replied:
My feeling that this needs ti be little different: For every source or sink named, say, "xyz": You create a dataset "xyz.external". The properties of the dataset contain a list of sources and sinks that use that name. That way, multiple souurces/sinks can share the same name (and hence ext dataset) and you can still track them back to the actual sources/sinks that they represent. 

The current code would only record the first one with its configuration, but the second time, because the dataset already exists, the configuration will not be stored.

poornachandra replied:
We're assuming here that sources/sinks that have the same name will have the same general properties. They can have runtime arguments that are different across runs. Hence it is okay to store the properties only once. 

That being said, today we don't have a way to way to distinguish properties from runtime arguments for a source or a sink. Both properties and arguments are combined together as configuration. Hence we'll end up storing properties + (runtime arguments of one source/sink) as the dataset properties.

I'll file a JIRA to address this in a later release.

poornachandra commented:
@anew I have replied to your comments, I'll work on addressing them. Can you take a look at the replies when you get a chance?

poornachandra commented:
@anew I have addressed the comments, please take a look.

One thing I noticed is that we don't support both naming and aliasing of outputs in `BatchSinkContext` and `MapReduceContext` like we do for inputs. 
* Datasets can only have names, no aliases. Hence we cannot use same the Dataset instance with different runtime arguments as multiple outputs in a batch program.
* Output format provider can take a name, but I think the name is used as an alias today. As a result, in this PR while creating external datasets for an output format provider, we would end up creating multiple dataset instances for each alias.

We  would need an Output class similar to Input to resolve this. Adding an Output class leads to a major API change. Can you think of any other clean fix that does not need major API change for this release?

poornachandra commented:
Also I think we can use the name of the input/output provider to check whether it is already a Dataset or not. If a Dataset with the same name exists, then we will use the existing Dataset with the provider configuration runtime arguments as the input/output for the batch job. 

What do you think of the above approach? This means we'll end up treating the name of providers in the same way as we treat Dataset instance names, which would be a change from the current behavior.

anew commented:
@poornachandra about your previous comment - adding an Output class - I remember suggesting that to @anwar6953 when we introduced the Input class - Ali, do you remember why we did not do that?

anew commented:
@poorna, regarding your last comment. I am not sure that would be correct. Would it mean that if a dataset "purchases" exists, and now someone creates a pipeline that has DB source, and he names that source purchases - would we record lineage for the existing dataset? Even though that dataset may not be involved in the pipeline at all...

anwar6953 commented:
@anew
Regarding your first comment:
We added the `Input` class because there are many different types of inputs we wanted to encapsulate in one object (So we dont have a large number of APIs, each taking in a different set of parameters).
The output had only a few types, so we didn't have this need.

anew commented:
@anwar6953 how do you feel about changing the way we specify outputs to be similar to inputs, using an Output wrapper. It is very short before the planned release, is there high risk associated with that?

anwar6953 commented:
The changes are really small.
I updated the JIRA with a summary of the updates: https://issues.cask.co/browse/CDAP-5651

It does make simple usage more complicated:
```java
context.addOutput("fooDataset");
// now becomes
context.addOutput(Output.ofDataset("fooDataset"));
```

What's the alternative to fulfill the need of this PR, though?

poornachandra commented:
@anew I have addressed comments, please take a look.

anew commented on cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/co/cask/cdap/etl/api/batch/BatchSinkContext.java:
```diff
@@ -48,11 +49,20 @@
+   * @deprecated Deprecated since 3.4.0.
+   *             Use {@link #addOutput(Output)} instead
    */
   void addOutput(String outputName, OutputFormatProvider outputFormatProvider);
```
@Deprecated

poornachandra replied:
Done

anew commented on cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/co/cask/cdap/etl/api/batch/BatchSinkContext.java:
```diff
@@ -28,15 +29,15 @@
    *
    * @param datasetName the name of the output dataset
    */
   void addOutput(String datasetName);
```
With addOutput(Output) is this still needed?

poornachandra replied:
Deprecated

anew commented on cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/co/cask/cdap/etl/api/batch/BatchSinkContext.java:
```diff
@@ -48,11 +49,20 @@
   void addOutput(String datasetName, Map<String, String> arguments);
```
same here?

poornachandra replied:
Deprecated

anew commented on cdap-app-templates/cdap-etl/cdap-etl-batch/src/main/java/co/cask/cdap/etl/batch/spark/SparkBatchSinkContext.java:
```diff
@@ -51,4 +53,10 @@ public void addOutput(String datasetName, Map<String, String> arguments) {
+
+  @Override
+  public void addOutput(Output output) {
+    Output trackableOutput = ExternalDatasets.makeTrackable(sparkContext.getAdmin(), output);
```
for the other 3 addOutput() methods, you don't need to make it trackable? What are those needed for? Just backward-compat?

poornachandra replied:
Yes - for backwards compatibility. I will mark them as deprecated.

anew replied:
Do we need to deprecate? Don't plugins specify explicitly what versions of the ETL apps they are compatible with? I believe that in 3.4 we are breaking compat anyway.

poornachandra replied:
I deprecated the methods, we can remove them in a later release.

anew commented on cdap-app-templates/cdap-etl/cdap-etl-batch/src/main/java/co/cask/cdap/etl/batch/spark/SparkBatchSinkFactory.java:
```diff
@@ -82,22 +83,47 @@ private SparkBatchSinkFactory(Map<String, OutputFormatProvider> providers,
+  private void addOutput(String stageName, String alias,
+                         BasicOutputFormatProvider outputFormatProvider) {
+    if (outputFormatProviders.containsKey(alias) || datasetInfos.containsKey(alias)) {
+      throw new IllegalArgumentException("OutputFormatProvider with name " + alias + " is already defined. " +
```
I think the message is very confusing, it exposes implementation, too. If the user calls addOutput(Output.ofDataset("x").alias("y") - assuming that it how it is supposed to be used? - but an output with alias "y" alsready exists, then the message is "OutputFormatProvider with name y is already defined. Cannot add multiple OutputFormatProviders with the same name". Even though what is clashing here is the alias, not the name. Also the message should just be in terms of "output" and perhaps it should mention the existing output that has the same alias.

poornachandra replied:
I just copied existing message. I agree it is confusing, I will update it.

anew commented on cdap-app-templates/cdap-etl/cdap-etl-batch/src/main/java/co/cask/cdap/etl/batch/spark/SparkBatchSinkFactory.java:
```diff
@@ -82,22 +83,47 @@ private SparkBatchSinkFactory(Map<String, OutputFormatProvider> providers,
+
+  private void addOutput(String stageName, String datasetName, String alias, Map<String, String> datasetArgs) {
+    if (outputFormatProviders.containsKey(alias) || datasetInfos.containsKey(alias)) {
+      throw new IllegalArgumentException("Output dataset with name " + alias + " is already defined. " +
```
And same here about the message.

anew commented on cdap-app-templates/cdap-etl/cdap-etl-batch/src/main/java/co/cask/cdap/etl/batch/spark/SparkBatchSourceContext.java:
```diff
@@ -82,6 +84,12 @@ public void setInput(String datasetName, Dataset dataset) {
 
+  @Override
+  public void setInput(Input input) {
+    Input trackableInput = ExternalDatasets.makeTrackable(sparkContext.getAdmin(), input);
```
here too, we don't make it trackable for the other setInput() methods?

poornachandra replied:
The other methods are for backwards compatibility. Since the other methods do not have the name of the input provider we cannot make it trackable.

anew commented on cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/co/cask/cdap/etl/common/ExternalDatasets.java:
```diff
@@ -0,0 +1,127 @@
+      Map<String, String> arguments = new HashMap<>();
+      arguments.put("input.format.class", inputFormatProvider.getInputFormatClassName());
+      arguments.putAll(inputFormatConfiguration);
+      if (!admin.datasetExists(datasetName)) {
```
With this, is is possible to have one pipeline writing to an external data store as a Sink, another reading from the same external store as it Source, and have them represented by the same external dataset? For example, pipeline A writing to a Kafka topic, pipeline B reading from that topic, and the alias would be the topic name, say myTopic, so that they become the same external dataset?

It seems that in this case, you would create an externalDataset ext-myTopic that - assuming A is created first, has the required output arguments for the sink, but then when B is deployed, the external dataset already exists and you will not record the input arguments. So even though ext-myTopic implements InputFormatProvider, it will fail to work as one. 

But we do want to support that use case, right?

poornachandra replied:
The exists check is only for creating the external dataset. If the external dataset already exists then it is used. The input arguments are passed as runtime arguments to the external dataset during its instantiation. 

In the above case when B is deployed it sees that ext-myTopic exists, and uses that with the runtime arguments of input format provider.

anew replied:
I see, so the properties are populated (for lineage display purposes) when the first pipeline is created, but ignored at runtime. So the program that uses this externalDataset still must provide the input/output format class name and the runtime arguments again, every time it runs. 

That also means that programs other the pipelines that create these external datasets cannot use them. For example, I cannot use this dataset an input for MapReduce. I might be tempted to do that if I see this dataset in the list of datasets. But I will only get a failure at runtime, right?

poornachandra replied:
Every time you use external dataset you are expected to give the run time arguments. We are storing runtime arguments as properties since we cannot separately identify them. I have filed JIRA for this - https://issues.cask.co/browse/CDAP-5674

anew commented:
Hi @poornachandra I am done with my review, please see my comments

anew commented on cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/co/cask/cdap/etl/common/ExternalDatasets.java:
```diff
@@ -0,0 +1,127 @@
+
+    try {
+      // Create an external dataset for the input format for lineage tracking
+      String datasetName = "ext-" + inputName;
```
I think "ext-" is not sufficiently unique, I can easily see someone calling his dataset "ext-requests" or something like that. It is much less likely to clash with actual dataset names and also makes it much clearer where this comes from, if you prefix it "hydrator-"

anwar6953 replied:
someone might name a dataset hydrator-meta also, though, right?

anew replied:
yes, but a lot less likely.

anew replied:
what would be a better way to name it without the risk of clash?

anew commented on cdap-app-templates/cdap-etl/cdap-etl-batch/src/main/java/co/cask/cdap/etl/batch/spark/SparkBatchSinkFactory.java:
```diff
@@ -82,22 +83,47 @@ private SparkBatchSinkFactory(Map<String, OutputFormatProvider> providers,
-      throw new IllegalArgumentException("OutputFormatProvider with name " + outputName + " is already defined. " +
-                                           "Cannot add multiple OutputFormatProviders with the same name.");
+  void addOutput(String stageName, Output output) {
+    if (output instanceof Output.DatasetOutput) {
```
can you add some comments here?

poornachandra replied:
Done

poornachandra commented:
Also filed a JIRA for new abstraction to track external sources/sinks - https://issues.cask.co/browse/CDAP-5675

poornachandra commented:
@anew I have addressed comments, please take another look.

anew commented:
LGTM.

poornachandra commented:
Thanks for the review. I'll squash and merge.

poornachandra commented:
There is one flaky unrelated test failure `UsageHandlerTestRun.testWorkerUsage()` in http://builds.cask.co/browse/CDAP-DUT3961-15. The same test ran fine on my laptop, I'll merge.
